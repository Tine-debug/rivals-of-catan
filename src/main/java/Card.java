// Card.java
// Quick & dirty, Basic-set only, refactored to reduce duplication

import java.io.IOException;

public class Card implements Comparable<Card> {

    // ---------- Public fields (keep simple for the take-home) ----------
    public String name, theme, type, cost, oneOf;
    public String cardText;

    public Points points;

    public String placement;

    private Logic logic;

    // Regions track “stored” resources by rotating; here we model it as an int
    // (0..3)
    public int regionProduction = 0;
    // Regions use production die faces (1..6). 0 means “not a region” / unassigned.
    public int diceRoll = 0;

    // ---------- Global piles for the Basic set ----------
    public static Cardstacks stacks = Cardstacks.getInstance();

    // ---------- Construction ----------
    public Card() {
    }

    public Card(String name, String theme, String type,
            String germanName, String placement,
            String oneOf, String cost,
            String victoryPoints, String CP, String SP, String FP,
            String PP, String LP, String KP, String Requires,
            String cardText, String protectionOrRemoval) {
        this.name = name;
        this.theme = theme;
        this.type = type;
        this.placement = placement;
        this.oneOf = oneOf;
        this.cost = cost;
        this.points = new Points(victoryPoints, CP, SP, FP, PP, LP, KP);
        this.cardText = cardText;

        this.logic = LogicFactory.createLogic(placement, name, type);
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public int compareTo(Card o) {
        return this.name.compareToIgnoreCase(o.name);
    }

    public static void loadBasicCards(String jsonPath) throws IOException {
        Cardstacks.loadBasicCards(jsonPath);
    }

    public boolean applyEffect(Player active, Player other, int row, int col) {
        String nm = (this.name == null ? "" : this.name);
        System.out.println("ApplyEffect: " + nm + " at (" + row + "," + col + ")");
        return logic.applyEffect(active, other, row, col, this);
    }

    public String summarizePoints(){
        return points.summarizePoints();
    }

}
