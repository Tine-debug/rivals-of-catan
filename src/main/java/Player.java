
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;

public class Player {

    public Points points = new Points("2", null, null, null, null, null, null);

    public int tradeRate = 3; 
    public boolean isBot = false;

    public Set<String> flags = new HashSet<>();

    public Map<String, Integer> resources = new HashMap<>();

    public List<Card> hand = new ArrayList<>();

    public Principality principality;

    public int lastSettlementRow = -1, lastSettlementCol = -1;

    private final Scanner in = new Scanner(System.in);

    public Player() {
        String[] all = { "Brick", "Grain", "Lumber", "Wool", "Ore", "Gold", "Any" };
        principality = new Principality();
        resources = new java.util.HashMap<>();
        for (String r : all)
            resources.put(r, 0);
    }

    public void sendMessage(Object m) {
        System.out.println(m);
    }

    public String receiveMessage() {
        System.out.print("> ");
        return in.nextLine();
    }

    public Card getCard(int r, int c) {
        return principality.getCard(r, c);
    }

    public void placeCard(int r, int c, Card card) {
        principality.placeCard(r, c, card);
    }


    public int expandAfterEdgeBuild(int col) {
        return principality.expandAfterEdgeBuild(col);
    }


    public boolean hasInPrincipality(String name) {
        return principality.hasInPrincipality(name);
    }

    public String printPrincipality() {
        return principality.printPrincipality(this);
    }


    public String printHand() {
        StringBuilder sb = new StringBuilder();
        sb.append("Hand (").append(hand.size()).append("):\n");
        for (int i = 0; i < hand.size(); i++) {
            Card c = hand.get(i);
            if (c == null)
                continue;
            String cost = (c.cost == null || c.cost.isBlank()) ? "-" : c.cost;
            String pts = c.summarizePoints();
            sb.append("  [").append(i).append("] ")
                    .append(c.name == null ? "Unknown" : c.name)
                    .append("   {cost: ").append(cost).append("} ")
                    .append(pts.isEmpty() ? "" : pts)
                    .append("\n").append(c.cardText == null ? "" : "\t" + c.cardText + "\n");
        }
        return sb.toString();
    }

  
    public boolean hasTradeTokenAgainst(Player opp) {
        if (this.points.commercePoints < 3) return false;
        if (opp == null) return true;
        return (this.points.commercePoints > opp.points.commercePoints);
    }

    public boolean hasStrengthTokenAgainst(Player opp) {
        if (this.points.strengthPoints < 3) return false;
        if (opp == null) return true;
        return (this.points.strengthPoints > opp.points.strengthPoints);
    }

    // Final score used for win check: base VP + 1 per advantage token against
    // opponent
    public int currentScoreAgainst(Player opp) {
        int score = this.points.victoryPoints;
        if (hasTradeTokenAgainst(opp))
            score += 1;
        if (hasStrengthTokenAgainst(opp))
            score += 1;
        return score;
    }


    // ------------- Resources (per-region, not pooled) -------------

    // Map a resource name to its Region card name
    private String resourceToRegion(String type) {
        if (type == null)
            return null;
        String t = type.trim().toLowerCase();
        switch (t) {
            case "brick":
                return "Hill";
            case "grain":
                return "Field";
            case "lumber":
                return "Forest";
            case "wool":
                return "Pasture";
            case "ore":
                return "Mountain";
            case "gold":
                return "Gold Field";
            case "any":
                return "Any";
            default:
                return null;
        }
    }



    public int totalAllResources() {
        return principality.totalAllResources();
    }

    // Count stored resources of a specific resource type across the board
    public int getResourceCount(String type) {
        return principality.getResourceCount(type);
    }

    // Gain 1 resource of a type: add to the matching region with the LOWEST stock
    // (<3)
    // If "Any", ask the player which resource to take.
    public void gainResource(String type) {
        String t = type;
        if (t == null || t.equalsIgnoreCase("Any")) {
            sendMessage("PROMPT: Choose resource to gain (Brick/Grain/Lumber/Wool/Ore/Gold):");
            t = receiveMessage();
        }
        String regionName = resourceToRegion(t);
        if (regionName == null || "Any".equals(regionName)) {
            sendMessage("Unknown resource '" + t + "'. Ignored.");
            return;
        }

        java.util.List<Card> regs = principality.findRegions(regionName);
        if (regs.isEmpty()) {
            sendMessage("No region for resource " + t + " is present.");
            return;
        }

        Card best = null;
        int bestVal = Integer.MAX_VALUE;
        for (Card r : regs) {
            int v = Math.max(0, Math.min(3, r.regionProduction));
            if (v < bestVal) {
                bestVal = v;
                best = r;
            }
        }
        if (best != null && best.regionProduction < 3) {
            best.regionProduction += 1;
        } else {
            sendMessage("No storage space on any " + regionName + " (already 3/3).");
        }
    }

 
    public boolean removeResource(String type, int n) {
        return principality.removeResource(type, n);
    }


    public void setResourceCount(String type, int n) {
        principality.setResourceCount(type, n);
    }

    // ------------- Hand -------------
    public int handSize() {
        return hand.size();
    }

    public void addToHand(Card c) {
        hand.add(c);
    }

    public Card removeFromHandByName(String nm) {
        for (int i = 0; i < hand.size(); i++) {
            Card c = hand.get(i);
            if (c != null && c.name != null && c.name.equalsIgnoreCase(nm)) {
                return hand.remove(i);
            }
        }
        return null;
    }

    // ------------- Cheap prompts used by Server -------------
    public String chooseResource() {
        sendMessage("PROMPT: Choose resource:");
        return receiveMessage();
    }
}