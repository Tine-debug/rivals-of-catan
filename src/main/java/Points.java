
public class Points {

    public int skillPoints = 0;
    public int strengthPoints = 0;
    public int victoryPoints = 0;
    public int progressPoints = 0;
    public int commercePoints = 0;
    public int sailPoints = 0;
    public int canonPoints = 0;

    public Points() {

    }

//String victoryPoints, String CP, String SP, String FP, String PP, String LP, String KP
    public Points(String victoryPoints, String CP, String SP, String FP, String PP, String LP, String KP) {
        this.victoryPoints = asInt(victoryPoints, 0);
        this.commercePoints = asInt(CP, 0);
        this.strengthPoints = asInt(SP, 0);
        this.skillPoints = asInt(FP, 0);
        this.sailPoints = asInt(LP, 0);
        this.canonPoints = asInt(KP, 0);

    }

    public static Points addPoints(Points points1, Points points2) {
        Points result = new Points();

        result.victoryPoints = points1.victoryPoints + points2.victoryPoints;
        result.commercePoints = points1.commercePoints + points2.commercePoints;
        result.strengthPoints = points1.strengthPoints + points2.strengthPoints;
        result.skillPoints = points1.skillPoints + points2.skillPoints;
        result.sailPoints = points1.sailPoints + points2.sailPoints;
        result.canonPoints = points1.canonPoints + points1.canonPoints;

        return result;

    }

    private int asInt(String s, int def) {
        try {
            if (s == null) {
                return def;
            }
            return Integer.parseInt(s.trim());
        } catch (Exception e) {
            return def;
        }
    }

    public String summarizePoints() {
        StringBuilder t = new StringBuilder();
        if (victoryPoints > 0 || commercePoints > 0 || progressPoints > 0 || strengthPoints > 0
                || skillPoints > 0) {
            t.append("[");
            if (victoryPoints > 0) {
                t.append("VP").append(victoryPoints).append(" ");
            }
            if (commercePoints > 0) {
                t.append("CP").append(commercePoints).append(" ");
            }
            if (strengthPoints > 0) {
                t.append("SP").append(strengthPoints).append(" ");
            }
            if (skillPoints > 0) {
                t.append("FP").append(skillPoints).append(" ");
            }
            if (progressPoints > 0) {
                t.append("PP").append(progressPoints).append(" ");
            }
            if (t.charAt(t.length() - 1) == ' ') {
                t.deleteCharAt(t.length() - 1);
            }
            t.append("]");
        }
        return t.toString();
    }
}
