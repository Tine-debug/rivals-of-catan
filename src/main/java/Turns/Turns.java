package Turns;

import Player.Player;
import Player.Broadcast;
import Card.Cardstacks;
import Card.Card;

import Turns.Phases.*;

import java.util.List;
import java.util.Random;

public class Turns {

    public Turns(List<Player> players) {
        this.players = players;
    }

    private final Broadcast broadcast = Broadcast.getInstance();
    private final Cardstacks stacks = Cardstacks.getInstance();

    private final Random rng = new Random();
    List<Player> players;

    public int resolveOneTurn(int current, boolean random) {
        Player active = players.get(current);
        Player other = players.get((current + 1) % players.size());

        int eventFace;
        int prodFace;
        if (random) {
            eventFace = rollEventDie(active);
            prodFace = rollProductionDie(active);
        } else {
            eventFace = getNonRandomEventDie();
            prodFace = getNonRandomProductionDie(active);
        }

        if (eventFace == DiceEvents.getBrigandFace()) {
            DiceEvents.resolveEvent(eventFace, active, other);
            applyProduction(prodFace);
        } else {
            applyProduction(prodFace);
            DiceEvents.resolveEvent(eventFace, active, other);
        }

        for (int i = 0; i < players.size(); i++) {
            players.get(i).sendMessage("Opponent's board:");
            players.get(i).sendMessage(
                    "\t\t" + players.get((i + 1) % players.size()).printPrincipality().replace("\n", "\n\t\t"));
            players.get(i).sendMessage("Your board:");
            players.get(i).sendMessage(players.get(i).printPrincipality());
            players.get(i).sendMessage("Your hand:");
            players.get(i).sendMessage(players.get(i).printHand());
        }

        actionphase.resolvePhase(active, other);

        replenish(active);

        ExchangePhase.resolvePhase(active);

        if (checkWinEndOfTurn(active, other)) {
            return 3;
        }

        current = (current + 1) % players.size();

        return current;

    }

    protected int rollEventDie(Player active) {
        int face = 1 + rng.nextInt(6);
        broadcast.broadcast("[EventDie] -> " + face);
        return face;
    }

    protected int rollProductionDie(Player active) {
        int face = 1 + rng.nextInt(6);
        if (active.flags.contains("BRIGITTA")) {
            active.sendMessage("PROMPT: Brigitta active -  choose production die [1-6]:");
            try {
                int forced = Integer.parseInt(active.receiveMessage().trim());
                if (forced >= 1 && forced <= 6) {
                    face = forced;
                }
            } catch (NumberFormatException ignored) {
            }
            active.flags.remove("BRIGITTA");
        }
        broadcast.broadcast("[ProductionDie] -> " + face);
        return face;
    }

    private void applyProduction(int face) {
        for (Player p : players) {
            boolean hasMarketplace = p.flags.contains("MARKETPLACE");
            int pMatches = countFaceRegions(p, face);
            int oppMatches = countFaceRegions(opponentOf(p), face);

            for (int r = 0; r < p.principality.principality.size(); r++) {
                List<Card> row = p.principality.principality.get(r);
                for (int c = 0; c < row.size(); c++) {
                    Card card = row.get(c);
                    if (card == null || !"Region".equalsIgnoreCase(card.type)) {
                        continue;
                    }
                    if (card.diceRoll != face) {
                        continue;
                    }
                    int inc = 1;
                    if (hasAdjacentBoosterForRegion(p, r, c)) {
                        inc += 1;
                    }

                    card.regionProduction = Math.min(3, card.regionProduction + inc);
                }
            }

            if (hasMarketplace && oppMatches > pMatches) {
                p.sendMessage("PROMPT: Marketplace - choose one resource produced on face " + face
                        + " to gain (e.g., Grain/Gold/Lumber):");
                String res = p.receiveMessage();
                p.gainResource(res);
            }
        }
    }

    public void replenish(Player p) {
        if (p.flags != null && p.flags.remove("NO_REPLENISH_ONCE")) {
            p.sendMessage("You cannot replenish your hand this turn (Fraternal Feuds).");
        } else {
            int handTarget = 3 + p.points.progressPoints;
            while (p.handSize() < handTarget) {
                p.sendMessage("PROMPT: Replenish - choose draw stack [1-4]:");
                int which = readInt(p.receiveMessage(), 1);
                stacks.drawCardfromStack(which, p);
            }
        }
    }

    private boolean checkWinEndOfTurn(Player active, Player other) {
        int score = active.currentScoreAgainst(other);
        if (score >= 7) {
            broadcast.broadcast("winner: Player " + players.indexOf(active)
                    + " wins with " + score + " VP (incl. advantage tokens)!");
            return true;
        }
        return false;
    }

    private int countFaceRegions(Player p, int face) {
        int n = 0;
        for (List<Card> row : p.principality.principality) {
            for (Card c : row) {
                if (c != null && "Region".equalsIgnoreCase(c.type) && c.diceRoll == face) {
                    n++;
                }
            }
        }
        return n;
    }

    private Player opponentOf(Player p) {
        return (p == players.get(0)) ? players.get(1) : players.get(0);
    }

    private boolean hasAdjacentBoosterForRegion(Player p, int rr, int cc) {
        Card region = p.getCard(rr, cc);
        if (region == null) {
            return false;
        }
        Card left = p.getCard(rr, cc - 1);
        Card right = p.getCard(rr, cc + 1);
        return isBoosting(left, region) || isBoosting(right, region);
    }

    private int readInt(String s, int def) {
        try {
            return Integer.parseInt(s.trim());
        } catch (NumberFormatException e) {
            return def;
        }
    }

    private boolean isBoosting(Card maybeBuilding, Card region) {
        if (maybeBuilding == null) {
            return false;
        }
        if (!"Building".equalsIgnoreCase(maybeBuilding.type)) {
            return false;
        }
        return buildingBoostsRegion(maybeBuilding.name, region.name);
    }

    public static boolean buildingBoostsRegion(String buildingName, String regionName) {
        if (buildingName == null || regionName == null) {
            return false;
        } else if (buildingName.equalsIgnoreCase("Iron Foundry") && regionName.equalsIgnoreCase("Mountain")) {
            return true;
        } else if (buildingName.equalsIgnoreCase("Grain Mill") && regionName.equalsIgnoreCase("Field")) {
            return true;
        } else if (buildingName.equalsIgnoreCase("Lumber Camp") && regionName.equalsIgnoreCase("Forest")) {
            return true;
        } else if (buildingName.equalsIgnoreCase("Brick Factory") && regionName.equalsIgnoreCase("Hill")) {
            return true;
        } else if (buildingName.equalsIgnoreCase("Weaverâ€™s Shop") && regionName.equalsIgnoreCase("Pasture")) {
            return true;
        } else if (buildingName.equalsIgnoreCase("Weaver's Shop") && regionName.equalsIgnoreCase("Pasture")) {
            return true;
        }
        return false;
    }

    int nonrandeventdie = 0;
    int nonrandproductiondie = 0;

    private int getNonRandomEventDie() {
        nonrandeventdie = (nonrandeventdie + 1) % 6;
        return nonrandeventdie + 1;
    }

    protected int getNonRandomProductionDie(Player active) {
        nonrandproductiondie = (nonrandproductiondie + 1) % 6;
        int face = 1 + nonrandproductiondie;
        if (active.flags.contains("BRIGITTA")) {
            active.sendMessage("PROMPT: Brigitta active -  choose production die [1-6]:");
            try {
                int forced = Integer.parseInt(active.receiveMessage().trim());
                if (forced >= 1 && forced <= 6) {
                    face = forced;
                }
            } catch (NumberFormatException ignored) {
            }
            active.flags.remove("BRIGITTA");
        }
        return face;
    }

}
