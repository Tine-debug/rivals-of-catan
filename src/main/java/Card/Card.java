package Card;

import java.io.IOException;
import Card.logic.*;
import Points.Points;
import Player.Player;

public class Card implements Comparable<Card> {

    // ---------- Public fields (keep simple for the take-home) ----------
    private String name, theme, type, cost;
    private boolean oneOf;
    private String cardText;

    private Points points;

    private String placement;

    private Logic logic;

    // Regions track “stored” resources by rotating; here we model it as an int
    // (0..3)
    private int regionProduction = 0;
    // Regions use production die faces (1..6). 0 means “not a region” / unassigned.
    private int diceRoll = 0;

    // ---------- Construction ----------
    public Card() {
    }

    protected Card(String name, String theme, String type,
        String placement, boolean oneOf, String cost,
        Points points, String cardText, Logic logic) {
        this.name = name;
        this.theme = theme;
        this.type = type;
        this.placement = placement;
        this.oneOf = oneOf;
        this.cost = cost;
        this.points = points;
        this.cardText = cardText;
        this.logic = logic;
    }


    //getters
    public String getTheme() {
        return theme;
    }

    public String getType() {
        return type;
    }

    public String getCost() {
        return cost;
    }

    public boolean getOneOf() {
        return oneOf;
    }

    public String getCardText() {
        return cardText;
    }

    public String getName() {
        return name;
    }

    public String getPlacement() {
        return placement;
    }

    public Points getPoints() {
        return points;
    }

    public int getRegionProduction() {
        return regionProduction;
    }

    public int getdiceRoll() {
        return diceRoll;
    }

    public void setPoints(Points newPoints) {
        points = newPoints;

    }

    public void setRegionProduction(int newcount) {
        if (newcount >= 0 && newcount <= 3) {
            regionProduction = newcount;
        }

    }

    public void setDiceRoll(int newdice) {
        if (newdice >= 0 && newdice <= 6) {
            diceRoll = newdice;
        }

    }

    @Override
    public String toString() {
        if (name != null) {
            return name;
        }
        return "?";
    }

    @Override
    public int compareTo(Card o) {
        return this.name.compareToIgnoreCase(o.name);
    }

    public boolean applyEffect(Player active, Player other, int row, int col) {
        String nm = (this.name == null ? "" : this.name);
        System.out.println("ApplyEffect: " + nm + " at (" + row + "," + col + ")");
        return logic.applyEffect(active, other, row, col, this);
    }

    public String summarizePoints() {
        return points.summarizePoints();
    }

}
