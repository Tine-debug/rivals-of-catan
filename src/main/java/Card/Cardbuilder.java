package Card;

import Points.Points;
import Card.logic.Logic;

public class Cardbuilder {

    private String name, theme, type, cost;
    private boolean oneOf;
    private String cardText;
    private Points points;
    private String placement;
    private Logic logic;



    public void name(String name) {
        this.name = name;
    }

    public void theme(String theme) {
        this.theme = theme;
    }

    public void type(String type) {
        this.type = type;
    }

    public void cost(String cost) {
        this.cost = cost;
    }

    public void oneOf(boolean oneOf){
        this.oneOf = oneOf;
    }

    public void oneOf(String oneOf){
        this.oneOf = (oneOf == null) ? false : (oneOf.equalsIgnoreCase("1x"));
    }

    public void cardText(String cardText){
        this.cardText = cardText;
    }

    public void points(Points points){
        this.points = points;
    }

    public void placement(String placement){
        this.placement = placement;
    }

    public void logic(Logic logic){
        this.logic = logic;
    }


    public Card build(){
        return new Card(this.name, this.theme, this.type, this.placement, 
        this.oneOf, this.cost, this.points, this.cardText, 
        this.logic);
    }

}
