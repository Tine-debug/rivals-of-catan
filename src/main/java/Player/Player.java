package Player;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;

import Points.*;
import Card.Card;

public class Player {



    private Points points;
    
    

    //private int tradeRate = 3;
    private boolean isBot = false;

    private Set<String> flags = new HashSet<>();

    private Map<String, Integer> resources = new HashMap<>();

    private final Hand hand = new Hand();

    private final Principality principality;

    private int lastSettlementRow = -1, lastSettlementCol = -1;

    private final Scanner in = new Scanner(System.in);

    public Player() {
        String[] all = {"Brick", "Grain", "Lumber", "Wool", "Ore", "Gold", "Any"};
        principality = new Principality();
        resources = new java.util.HashMap<>();
        for (String r : all) {
            resources.put(r, 0);
        }
        PointsBuilder pointsBuilder = new PointsBuilder();
        pointsBuilder.victoryPoints(2);
        this.points = pointsBuilder.build();
    }

    public void sendMessage(Object m) {
        System.out.println(m);
    }

    public String receiveMessage() {
        System.out.print("> ");
        return in.nextLine();
    }

    //getters
    public Set<String> getFlags(){
        return this.flags;
    }

    public Principality getPrincipality(){
        return principality;
    }

    public Hand getHand(){
        return hand;
    }

    public boolean getIsBot(){
        return isBot;
    }

    public int getLastSettlementRow(){
        return lastSettlementRow;
    }

    public int getLastSettlementCol(){
        return lastSettlementCol;
    }

    //setters
    public void setFlags(Set<String> newFlags){
        this.flags = newFlags;
    }

    public void setIsBot(boolean newvalue){
        this.isBot = newvalue;
    }

    public void setLastSettlementRow(int newvalue){
        this.lastSettlementRow = newvalue;
    }
    
    public void setLastSettlementCol(int newvalue){
        this.lastSettlementCol = newvalue;
    }

    

    //


    public void removeFlag(String toRemove){
        flags.remove(toRemove);
    }

    public Points getPoints(){
        return this.points;
    }

    public void setPoints(Points newValue){
        this.points = newValue;
    }



    public Card getCard(int r, int c) {
        return principality.getCard(r, c);
    }

    public void placeCard(int r, int c, Card card) {
        principality.placeCard(r, c, card);
    }

    public int expandAfterEdgeBuild(int col) {
        return principality.expandAfterEdgeBuild(col);
    }

    public boolean hasInPrincipality(String name) {
        return principality.hasInPrincipality(name);
    }

    public String printPrincipality() {
        return principality.printPrincipality(this);
    }

    public String printHand() {
        return hand.printHand();
    }

    public boolean hasTradeTokenAgainst(Player opp) {
        if (this.points.commercePoints < 3) {
            return false;
        }
        if (opp == null) {
            return true;
        }
        return (this.points.commercePoints > opp.points.commercePoints);
    }

    public boolean hasStrengthTokenAgainst(Player opp) {
        if (this.points.strengthPoints < 3) {
            return false;
        }
        if (opp == null) {
            return true;
        }
        return (this.points.strengthPoints > opp.points.strengthPoints);
    }

    // Final score used for win check: base VP + 1 per advantage token against
    // opponent
    public int currentScoreAgainst(Player opp) {
        int score = this.points.victoryPoints;
        if (hasTradeTokenAgainst(opp)) {
            score += 1;
        }
        if (hasStrengthTokenAgainst(opp)) {
            score += 1;
        }
        return score;
    }

    // ------------- Resources (per-region, not pooled) -------------
    // Map a resource name to its Region card name
    private String resourceToRegion(String type) {
        if (type == null) {
            return null;
        }
        String t = type.trim().toLowerCase();
        switch (t) {
            case "brick":
                return "Hill";
            case "grain":
                return "Field";
            case "lumber":
                return "Forest";
            case "wool":
                return "Pasture";
            case "ore":
                return "Mountain";
            case "gold":
                return "Gold Field";
            case "any":
                return "Any";
            default:
                return null;
        }
    }

    public int totalAllResources() {
        return principality.totalAllResources();
    }

    // Count stored resources of a specific resource type across the board
    public int getResourceCount(String type) {
        return principality.getResourceCount(type);
    }

    // Gain 1 resource of a type: add to the matching region with the LOWEST stock
    // (<3)
    // If "Any", ask the player which resource to take.
    public void gainResource(String type) {
        String t = type;
        if (t == null || t.equalsIgnoreCase("Any")) {
            sendMessage("PROMPT: Choose resource to gain (Brick/Grain/Lumber/Wool/Ore/Gold):");
            t = receiveMessage();
        }
        String regionName = resourceToRegion(t);
        if (regionName == null || "Any".equals(regionName)) {
            sendMessage("Unknown resource '" + t + "'. Ignored.");
            return;
        }

        java.util.List<Card> regs = principality.findRegions(regionName);
        if (regs.isEmpty()) {
            sendMessage("No region for resource " + t + " is present.");
            return;
        }

        Card best = null;
        int bestVal = Integer.MAX_VALUE;
        for (Card r : regs) {
            int v = Math.max(0, Math.min(3, r.getRegionProduction()));
            if (v < bestVal) {
                bestVal = v;
                best = r;
            }
        }
        if (best != null && best.getRegionProduction() < 3) {
            best.setRegionProduction(best.getRegionProduction() + 1);
        } else {
            sendMessage("No storage space on any " + regionName + " (already 3/3).");
        }
    }

    public boolean removeResource(String type, int n) {
        return principality.removeResource(type, n);
    }

    public void setResourceCount(String type, int n) {
        principality.setResourceCount(type, n);
    }

    // ------------- Hand -------------
    public int handSize() {
        return hand.handSize();
    }

    public void addToHand(Card c) {
        hand.addToHand(c);
    }

    public Card removeFromHandByName(String nm) {
        return hand.removeFromHandByName(nm);
    }

    // ------------- Cheap prompts used by Server -------------
    public String chooseResource() {
        sendMessage("PROMPT: Choose resource:");
        return receiveMessage();
    }

    public void refundCost(String cost) {
        if (cost == null || cost.isBlank()) {
            return;
        }
        Map<String, Integer> need = parseCost(cost);
        for (var e : need.entrySet()) {
            setResourceCount(e.getKey(), getResourceCount(e.getKey()) + e.getValue());
        }
    }

    public Map<String, Integer> parseCost(String cost) {
        Map<String, Integer> m = new HashMap<>();
        if (cost == null) {
            return m;
        }

        // Accept strings like "LW", "AA", with optional spaces or separators ("L,W", "A
        // A")
        for (int i = 0; i < cost.length(); i++) {
            char ch = cost.charAt(i);
            if (Character.isWhitespace(ch) || ch == ',' || ch == ';' || ch == '+') {
                continue;
            }
            String res = letterToResource(ch);
            if (res != null) {
                m.put(res, m.getOrDefault(res, 0) + 1);
            }
            // else: silently ignore unknown chars
        }
        return m;
    }

    public boolean payCost(String cost) {
        if (cost == null || cost.isBlank()) {
            return true;
        }
        // Cost like "1 Brick, 1 Grain, 1 Wool, 1 Lumber" etc.
        Map<String, Integer> need = parseCost(cost);
        for (var e : need.entrySet()) {
            if (getResourceCount(e.getKey()) < e.getValue()) {
                return false;
            }
        }
        for (var e : need.entrySet()) {
            removeResource(e.getKey(), e.getValue());
        }
        return true;
    }

    private String letterToResource(char ch) {
        switch (Character.toUpperCase(ch)) {
            case 'B':
                return "Brick";
            case 'G':
                return "Grain";
            case 'L':
                return "Lumber";
            case 'W':
                return "Wool";
            case 'O':
                return "Ore";
            case 'A':
                return "Gold";
            default:
                return null; // unknown / ignore
        }
    }


    public boolean hasStrengthAdvantage(Player other) {
        return this.points.strengthPoints >= 3 && this.points.strengthPoints > other.points.strengthPoints;
    }

}
