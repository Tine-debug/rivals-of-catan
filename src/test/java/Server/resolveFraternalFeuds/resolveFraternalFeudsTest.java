
import java.io.OutputStream;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Vector;
import java.util.stream.Stream;

import Card.*;
import Card.logic.*;
import Player.*;
import Turns.*;
import Turns.Cardevents.*;
import Turns.Phases.*;
import Points.*;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;


public class resolveFraternalFeudsTest{

    Cardstacks stacks = Cardstacks.getInstance();
    static private PrintStream originalOut;
    static private PrintStream originalErr;
    static CardEvent event = new fraternalfeudsEvent();

    @BeforeAll
    static void suppressOutput() {
        originalOut = System.out;
        originalErr = System.err;
        System.setOut(new PrintStream(OutputStream.nullOutputStream()));
        System.setErr(new PrintStream(OutputStream.nullOutputStream()));
    }

    @AfterAll
    static void restoreOutput() {
        System.setOut(originalOut);
        System.setErr(originalErr);
    }

     int center = 2;
int[][] regionDice = { { 2, 1, 6, 3, 4, 5 }, { 3, 4, 5, 2, 1, 6 } };

    Server server = new Server();
    MockPlayer player1;
    MockPlayer player2;

    public void setupPlayers(){
        try {
                stacks.sortIntoPiles(false, "cards.json");
        } catch (Exception e) {
        }
        player1 = new MockPlayer();
        player2 = new MockPlayer();
        PointsBuilder pointsBuilder = new PointsBuilder();
        pointsBuilder.victoryPoints(4);
        pointsBuilder.canonPoints(4);
        pointsBuilder.commercePoints(4);
        pointsBuilder.progressPoints(4);
        pointsBuilder.strengthPoints(4);
        pointsBuilder.skillPoints(4);
        pointsBuilder.sailPoints(4);
        player2.points = pointsBuilder.build();
        server.players.add(player1);
        server.players.add(player2);

        server.pricipalityinitoneplayer(player1, regionDice, center, 0);
        server.pricipalityinitoneplayer(player2, regionDice, center, 1);
        player2.messages.add(new String[]{"1 2"});
    }

    @Test
    public void fraternalFreudsTest(){
        setupPlayers();
        server.replenish(player1);
        server.replenish(player2);
        String result = "\n";
        result += player1.hand.printHand() + "\n";
        result += player2.hand.printHand() + "\n";
        event.resolve(player1, player2);
        result += player1.hand.printHand() + "\n";
        result += player2.hand.printHand() + "\n";

        Approvals.verify(result);

    }
    @Test
    public void fraternalFreudsemptyhandTest(){
        setupPlayers();
        String result = "\n";
        result += player1.hand.printHand() + "\n";
        result += player2.hand.printHand() + "\n";
        event.resolve(player1, player2);
        result += player1.hand.printHand() + "\n";
        result += player2.hand.printHand() + "\n";

        Approvals.verify(result);

    }

    @Test
    public void fraternalFreudsemptyNoAdvantageTest(){
        setupPlayers();
        String result = "\n";
        server.replenish(player1);
        server.replenish(player2);
        player2.points = new Points();
        result += player1.hand.printHand() + "\n";
        result += player2.hand.printHand() + "\n";
        event.resolve(player1, player2);
        result += player1.hand.printHand() + "\n";
        result += player2.hand.printHand() + "\n";

        Approvals.verify(result);

    }


           




}