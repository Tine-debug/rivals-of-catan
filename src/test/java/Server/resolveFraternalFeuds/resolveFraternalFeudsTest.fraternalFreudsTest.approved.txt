
Hand (3):
  [0] Abbey   {cost: BGO} 
	Progress is not the only thing here; you also get red wine and lots of dark beer.
  [1] Abbey   {cost: BGO} 
	Progress is not the only thing here; you also get red wine and lots of dark beer.
  [2] Austin   {cost: GO} [SP1 FP2]
	If you hit my left cheek, don’t even think you’ll have time to hit the right one too.

Hand (3):
  [0] Brick Factory   {cost: BO} 
	Doubles the brick production of the neighboring hills.
  [1] Brick Ship   {cost: LW} [CP1]
	During your turn, you may trade 2 brick for any 1 other resource as often as you wish.
  [2] Brigitta, the Wise Woman   {cost: -} 
	Play this card before rolling the dice. Choose the result of the production die roll.

Hand (1):
  [0] Abbey   {cost: BGO} 
	Progress is not the only thing here; you also get red wine and lots of dark beer.

Hand (3):
  [0] Brick Factory   {cost: BO} 
	Doubles the brick production of the neighboring hills.
  [1] Brick Ship   {cost: LW} [CP1]
	During your turn, you may trade 2 brick for any 1 other resource as often as you wish.
  [2] Brigitta, the Wise Woman   {cost: -} 
	Play this card before rolling the dice. Choose the result of the production die roll.

